rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+
+    ```
+    {
+      features {
+        "my_feature": {
+          bytes_list {
+          }
+        }
+    }
+    ```
+
+*   a feature is not present
+
+    ```
+    {
+      features {
+      }
+    }
+    ```
+
+    or
+
+    ```
+    {
+      features {
+        "my_feature": {}  # none of the oneof is set
+      }
+    }
+    ```
+
+Each feature can be encoded as:
+
+```
+[FixedSize]ListArray<int64|float32|binary>
+```
+
+Then, the feature value in case a) is encoded as an empty sub-list, while the
+feature value in case b) is encoded as null.
+
+If we know that all the lists in a `ListArray` are of equal length (from the
+schema of the data, see below sections), `FixedSizeListArray` can be used to
+obviate the `O(N)` space overhead for lengths of lists.
+
+### tf.SequenceExample
+
+[Conformant](https://github.com/tensorflow/tensorflow/blob/abfba15cd9734cec7ecd3d0661b146fc251c842d/tensorflow/core/example/example.proto#L184)
+tf.SequenceExamples are assumed. I/O + parsing should throw an error upon
+non-conformant instances.
+
+A context feature will be encoded similarly to a feature in tf.Example. A
+sequence feature will be encoded as:
+
+```
+[FixedSize]ListArray<[FixedSize]ListArray<int64|float32|binary>>
+```
+
+To avoid name conflicts with context features, all the sequence features can be
+grouped into one `StructArray`:
+
+```
+StructArray<{'sequence_feature1': ListArray<ListArray<int64|float32|binary>>, ...}>
+```
+
+### Structured data (e.g. Protocol Buffers / Apache Avro)
+
+A batch of structured records can be encoded as follows:
+
+*   Each direct leaf field of the structure can be encoded similarly to
+    tf.Example. (`ListArray` of primitive types).
+*   Each sub-message can be encoded as:
+
+    ```
+    ListArray<StructArray<recursion...>>>
+    ```
+
+## Arrow to TF Feedable conversion
+
+### TensorRepresentation
+
+One or more Arrow columns can potentially be converted to multiple types of TF
+feedables.
+
+For example, a `ListArray<int64>` can be converted to:
+
+*   a Tensor, if given a default value to pad
+*   a SparseTensor to represent a ragged array
+*   a RaggedTensor
+
+The choice depends on user’s intents, which currently is
+[implicitly](https://github.com/tensorflow/transform/blob/11afcff467f779ba6163686395582e69603987d1/tensorflow_transform/tf_metadata/schema_utils.py#L172)
+expressed in the pipeline schema.
+
+We propose to create a new [TFMD](https://github.com/tensorflow/metadata)
+(TensorFlow MetaData) Proto, `TensorRepresentation` to carry those intents implicitly:
+
+```protobuf
+message TensorRepresentation {
+  oneof {
+    DenseTensor { … }  // column_name, dtype, shape, default_value
+    VarLenSparseTensor { … }  // column_name, dtype
+    SparseTensor { }  // dtype, value_column_name, indice_column_names
+    VarLenRaggedTensor { … } // dtype
+    RaggedTensor { } // dtype, value_column_name, row_partition_column_names, ...
+    StructuredTensor { } // column_names
+  }
```

Comment:
Will new tensor-like types need to be updated manually here? It seems like it will be easy to miss types here over time.

Reply:
New tensor-like types that TFXIO will support (i.e. the types the user can get from the dataset returned by TFXIO.TFDataset()) need to be updated manually here. The support needs to be built, so it's unlikely that we "miss" something here, but we lag behind in supporting them.

Note that this (TFX support lagging behind for various tensor-like types) has always been the case. Currently TFX (mainly TFT) only supports dense tensors (with fixed shape), var len (ragged) sparse tensors and (poorly) sparse tensors (through COO encoding). We are already missing RaggedTensors whose parsing support has been added recently.

What's worse is, what will be encoded in TensorRepresentation currently is encoded in various fields in the TFMD schema, for example:

Schema::Feature.value_count.min == Schema::Feature.value_count.max ==> Dense Tensor

Schema::Feature.value_count.min != Schema::Feature.value_count.max || Feature.value_count.min == 0 ==> var len (ragged) SparseTensor

Schema::SparseFeature specified ==> (COO) SparseTensor